reframe/frontend/executors/policies.py

```diff
@@ -224,19 +262,49 @@ def runcase(self, case):
             self._failall(e)
             raise
 
+    def _cleanup_all(self):
+        for task in self._retired_tasks:
+            if task.ref_count == 0:
+                task.cleanup(not self.keep_stage_files)
+                self._retired_tasks.remove(task)
+
     def _poll_tasks(self):
         '''Update the counts of running checks per partition.'''
         getlogger().debug('updating counts for running test cases')
         getlogger().debug('polling %s task(s)' % len(self._running_tasks))
         for t in self._running_tasks:
             t.poll()
 
+    def _poll_waiting(self):
```

Comment:
Don't call this `_poll_xxx`, because you are not actually polling the tasks. I'd call it `_notify_waiting()`, since it notifies all the tasks waiting on dependencies.